### ISSUE TITLE
Fundeps: shorthands for common idioms, esp with Overlaps

### ISSUE THREAD

**Issue by AntC2** [ghc-proposals/ghc-proposals#219]
[I'll be building this up over the next few days. Remove this message when I'm done.]

### **Motivation**

There are common idioms for writing FunDep code. Furthermore if combined with Overlaps, the 'textbook' examples don't work reliably, so needing rather verbose idioms to get dependable behaviour. Here I'm thinking out loud about shorthands. This isn't introducing any new functionality or semantics/nothing to see here/move along please.

(These thoughts are prompted mostly by reviewing GHC tickets of unreliable/bogus behaviour and the work-rounds; also somewhat by looking at [the 'Habit' flavour of Haskell](https://hasp.cs.pdx.edu//habit-report-Nov2010.pdf) -- I can see what they're aiming for, but I dislike the specifics.)

### **1. Class decls**

By far the most common FunDeps have:

* a single FunDep;
* the target type parameter as the rightmost in the class decl;
* the Fundep is 'full' -- that is, all the other type params are sources.

And that structure is natural: the class acting as a type-level function.

    class C a b c  | a b -> c  where ...

suggested shorthand

    class C a b -> c  where ...

**Discussion/possible complications/objections**

* _If you want a type-level function, use a Type Family_

  Type Classes serve three purposes; whereas Type Families serve only one of those: a) a predicate on types; b) an overloading mechanism for methods; c) a mechanism for type improvement. Type Families sometimes work well in combination with type classes, sometimes not. This shorthand is for where not.

* _What if I want other FunDeps as well?_

  I'm not proposing taking away the longer-hand current syntax. Would be nice (but not essential) to combine them:

        class C a b -> c  | c b -> a  where ...

* _Bikeshed about lexical syntax_

  Habit uses `=` where I've used `->`. I had a strong reaction against that: `=` is what we use in equations, but this is a decl. So I think it's appropriate to use type-flavoured lexemes. (Type Families use `=` in the instances/equations, not the TF decl.) `->` for the target type fits with thinking of FunDeps as type-level functions; also directly corresponds to `->`in a fully written out FunDep. (The arrow comes from Relational Database theory of FunDeps.)

**Comment by goldfirere** [comment]
This doesn't conform to the usual template for proposals. I like to zoom straight to the "Proposed Change Specification", where I expect a terse, minimal description of what is proposed. Is there such a specification here?

**Comment by AntC2** [comment]
@goldfirere: double standards. The 'ViaApplications' so-called proposal has a single bullet in its "Proposed Change Specification". That is terse to the point of incomprehensibility. Is "there such a specification" there? Have you made such an unsympathetic, not to say peremptory, remark?

1. This here is not a proposal, which is why I've used the 'Issues' part of GitHub. As I understand it, 'Issues' are for more exploratory discussions/for "thinking out loud" (as I explicitly say)/for testing opinion of the collective.

2. This here applies exclusively for FunDeps. Going by all my experiences over the past few years, nobody at GHC HQ is paying any attention to FunDeps. Then were this to be a proposal, its chances of making its way into GHC are nil. Submitting to all the GitHub bureaucracy/formatting would not improve those chances. Why would I bother?

I'd appreciate some sympathetic feedback.

**Comment by nomeata** [comment]
@AntC2 you are right that using an issue for discussion is perfectly fine, and there is no requirement to follow the template for proposals. But this is not something that has been done a lot yet, so please give Richard the benefit of doubt that he was a bit confused, and voiced his confusion (admittedly in a rather terse tone).

```
instance (c ~ Char) C Int Bool c  where ...            -- (B)
```
this misses a `=>`, right?


> ```
> instance C ty1 ty2 ~ty3  where ...
> ```
> shorthand for
> ```
> instance (c ~ ty3) => C ty1 ty2 c  where ...
> ```

It is just a coincidence that this has both syntactic and semantic analogies with lazy patterns in terms?… ah, you mention that later as well.

Unfortunately, I have no personal experience with either programming with FunDeps nor with the corresponding code in GHC, so I don’t have an opinion here yet. I recommend you bang the drums a bit (mailing lists, slack, discourse, reddit, twitter) and see if other FunDep users can voice their opinion here.

**Comment by AntC2** [comment]
Thanks @nomeata . And thanks, yes fixed. And yes the `~` irrefutable type improvement analogue to irrefutable patterns in terms is conscious. 

Note the note at the top (not yet removed). I suspect there are significant numbers of users of FunDeps flying under the radar, who've stopped paying attention to GHC language evolution because it's not delivering any benefits for them; as opposed to the succinct yet readable, elegant language that's been the heart of Haskell. They've learnt the longhand idioms I'm talking about as the disciplined way that works/perhaps even regard them as inevitable. Then yes I'm mulling over how to canvass them. (Discussing language improvements was so much more vibrant when I started Haskelling.)

**Comment by jvanbruegge** [comment]
Just my two cents, but I am not a fan of syntax sugar if it does not bring a lot to the table. From what I can tell, this proposal is simply saving a few keystrokes. Is there anything else this brings to the table? I am a heavy user of fundeps myself, but I simply don't see much value here

**Comment by AntC2** [comment]
Thanks @jvanbruegge, it's always valuable to get feedback from heavy users.

Deliberately this is only considering syntactic shorthands. There's lots of other tickets and proposals for semantic improvements.

> Is there anything else this brings to the table?

I suspect many users are following the textbook idioms. Then they're writing code that is unreliable, and might give incoherent behaviour in non-obvious ways. If you're writing rugged code, then good on you, but it is considerably more verbose. I was trying to offer a middle ground. Then can I ask ...

Do your classes with FunDeps usually look like this?:

    class C a b c  | a b -> c  where ...

That is: a single FunDep; target the right hand type. [Note **]

Which of these do your instances usually look like?:

    instance C Int Bool Char  where ...                    -- (A)

    instance (c ~ Char) => C Int Bool c  where ...         -- (B)

If (A), do you understand that's unreliable? If (B), what did you do before there was `~` for type equality? Do the readers of your code find (B) harder to understand? Do newbies get scared by `UndecidableInstances`?

Note **: heavy users of Lenses seem to prefer FunDeps, but their classes/instances don't typically have a single FunDep:

    class Lensish s t a b  | s -> a, t -> b, s b -> t, t a -> s  where ...

The equivalent code using Associated Type/Families and superclass equality constraints requires a lot of boilerplate, so I'm not surprised people prefer FunDeps. With all those multi-directional FunDeps, you're pretty much forced to write the instances per style (A) above, with repeated tyvars in the instance head:

    instance ThirdofTuple (a1, a2, a3) (a1, a2, a3') a3 a3'  where ...

And with luck, if the client code gets its types out of position, all that will happen is a compile rejection. But that style is not avoiding the risk of incoherence. I'd rather do that with an irrefutable match

    instance ThirdofTuple (a1, a2, a3) (a1, a2, a3') ~a3 ~a3'  where ...

**Comment by goldfirere** [comment]
Mea culpa. (And I've been swamped this past week -- please do not place any semantic value on the delay of this response.) My comment was indeed phrased rudely. Furthermore, I had not noticed that this was an Issue, not a Pull Request. I agree that the more open-ended "proposal-building" makes much more sense in an Issue. The reason for my different treatment of the "via" posting is that I could get the nub of that proposal easily from reading the first few sentences; that experience did not replicate itself here. You might see that, later on, I requested clarification on that other proposal, too.

In any case, I apologize for my post, and I will try to be gentler in the future.

**Comment by AntC2** [comment]
Thanks @goldfirere 

> ... get the nub ... easily from reading the first few sentences; that experience did not replicate itself here.

Really, the first para (4 sentences) here doesn't give the nub? Or at least puts the reader in a position to decide whether they're interested in the idea.

I've now added a Summary/Proposed Change Specification. (That's a series of code examples, which I think is an adequate basis for "proposal-building".) And taken away the 'in progress' disclaimer.

Re the "via" proposal, I wasn't aiming to denigrate it; I understand it also to be "proposal-building". I (still after all those discussions) see no 'nub'. I see only a forest of `@` signs, and keyword `via`; oh and a "(handwave)". I've no idea whether it's something I might find useful. I suppose it's the sort of thing that the people who know what it's talking about know what it's talking about. Is that the basis on which the Steering Committee evaluates proposals?

**Comment by goldfirere** [comment]
Thanks for adding that summary -- I now see what you're getting at.

I do feel the need to address one point: 
> Is that the basis on which the Steering Committee evaluates proposals?

Not at all. I do hope that the committee has been consistent in requiring full specifications from proposals before being accepted. Naturally, we have to understand the proposals in order to debate them, but the committee has a range of participants who enjoy different parts of the Haskell language, and any specification would have to appeal broadly.

When I post on a proposal, I post as me, not as the committee. My opinions differ with that of others on the committee from time to time, as I would hope. I recognize that it may be confusing to have committee members actively participate before official submission to the committee, but it seems better to allow that than the alternative. In any case, please interpret my comments here as just coming from me, not as representative of anything larger. When this proposal is ready for submission to the committee, it can go through the regular process.
